server/vulnerabilities/cpe.go

```diff
@@ -139,67 +144,312 @@ func cleanAppName(appName string) string {
+	}
+
+	return translations, nil
+}
+
+func DownloadCPETranslations(vulnPath string, client *http.Client) error {
+	ghClient := github.NewClient(client)
+
+	opts := &github.RepositoryContentGetOptions{
+		Ref: "michal-6628-macos-vuln",
+	}
+	r, resp, err := ghClient.Repositories.DownloadContents(context.Background(), "fleetdm", "nvd", cpeTranslationsFilename, opts)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("github response non 200 status code: %d", resp.StatusCode)
+	}
+
+	tmpFile, err := os.CreateTemp(vulnPath, cpeTranslationsFilename)
+	if err != nil {
+		return fmt.Errorf("create temporary file: %w", err)
+	}
+	defer tmpFile.Close()
+
+	// Clean up tmp file if not moved
+	moved := false
+	defer func() {
+		if !moved {
+			os.Remove(tmpFile.Name())
+		}
+	}()
+
+	if _, err := io.Copy(tmpFile, r); err != nil {
+		return fmt.Errorf("write temporary file: %w", err)
+	}
+
+	if err := tmpFile.Close(); err != nil {
+		return fmt.Errorf("write and close temporary file: %w", err)
 	}
 
-	checkTargetSW := ""
-	args := []interface{}{onlyAlphaNumeric.ReplaceAllString(cleanAppName(software.Name), " ")}
-	if targetSW != "" {
-		checkTargetSW = " AND target_sw MATCH ?"
-		args = append(args, targetSW)
+	path := filepath.Join(vulnPath, cpeTranslationsFilename)
+
+	if err := os.Rename(tmpFile.Name(), path); err != nil {
+		return fmt.Errorf("rename temporary file: %w", err)
+	}
+
+	moved = true
+
+	return nil
+}
+
+// CPETranslations include special case translations for software that fail to match entries in the NVD CPE Dictionary
+// using the standard logic. This may be due to unexpected vendor or product names.
+//
+// Example:
+//   [
+//     {
+//       "match": {
+//         "bundle_identifier": ["com.1password.1password"]
+//       },
+//       "translation": {
+//         "product": ["1password"],
+//         "vendor": ["agilebits"]
+//       }
+//     }
+type CPETranslations []CPETranslationEntry
+
+func (c CPETranslations) Translate(s *fleet.Software) (CPETranslation, bool) {
+	for _, entry := range c {
+		if entry.Match.Matches(s) {
+			return entry.Translation, true
+		}
 	}
-	args = append(args, software.Version)
-
-	query := fmt.Sprintf(
-		`SELECT rowid, * FROM cpe WHERE rowid in (
-				  SELECT rowid FROM cpe_search WHERE title MATCH ?%s
-				) and version=? order by deprecated asc`,
-		checkTargetSW,
-	)
+
+	return CPETranslation{}, false
+}
+
+type CPETranslationEntry struct {
+	Match       CPETranslationMatch `json:"match"`
+	Translation CPETranslation      `json:"translation"`
+}
+
+// CPETranslationMatch represents match criteria for cpe translations.
+type CPETranslationMatch struct {
+	Name             []string `json:"name"`
+	BundleIdentifier []string `json:"bundle_identifier"`
+	Source           []string `json:"source"`
+}
+
+func (c CPETranslationMatch) Matches(s *fleet.Software) bool {
+	for _, name := range c.Name {
+		if name != s.Name {
```

Comment:
I thought each match value was meant to be combined using an `OR` operator? For example:
```
{
       "match": {
         "bundle_identifier": ["a", "b"]
       },
       "translation": {
...
       }
//     }
```
Will match if `bundle_identifier` is 'a' `OR` 'b'

Reply:
Ah, there was an issue in the logic I used here. Fixed it up and added a test.